services/winget/winget-version.service.js

```diff
@@ -0,0 +1,97 @@
+import Joi from 'joi'
+import gql from 'graphql-tag'
+import { latest, renderVersionBadge } from '../version.js'
+import { InvalidParameter, pathParam } from '../index.js'
+import { GithubAuthV4Service } from '../github/github-auth-service.js'
+import { transformErrors } from '../github/github-helpers.js'
+
+const schema = Joi.object({
+  data: Joi.object({
+    repository: Joi.object({
+      object: Joi.object({
+        entries: Joi.array().items(
+          Joi.object({
+            type: Joi.string().required(),
+            name: Joi.string().required(),
+          }),
+        ),
+      })
+        .allow(null)
+        .required(),
+    }).required(),
+  }).required(),
+}).required()
+
+export default class WingetVersion extends GithubAuthV4Service {
+  static category = 'version'
+
+  static route = {
+    base: 'winget/v',
+    pattern: ':name',
+  }
+
+  static openApi = {
+    '/winget/v/{name}': {
+      get: {
+        summary: 'WinGet Package Version',
+        description: 'WinGet Community Repository',
+        parameters: [
+          pathParam({
+            name: 'name',
+            example: 'Microsoft.WSL',
+          }),
+        ],
+      },
+    },
+  }
+
+  static defaultBadgeData = {
+    label: 'winget',
+  }
+
+  async fetch({ name }) {
+    const nameFirstLower = name[0].toLowerCase()
+    const nameSlashed = name.replaceAll('.', '/')
+    const path = `manifests/${nameFirstLower}/${nameSlashed}`
+    const expression = `HEAD:${path}`
+    return this._requestGraphql({
+      query: gql`
+        query RepoFiles($expression: String!) {
+          repository(owner: "microsoft", name: "winget-pkgs") {
+            object(expression: $expression) {
+              ... on Tree {
+                entries {
+                  type
+                  name
+                }
+              }
+            }
+          }
+        }
+      `,
+      variables: { expression },
+      schema,
+      transformErrors,
+    })
+  }
+
+  async handle({ name }) {
+    try {
+      const json = await this.fetch({ name })
+      if (json.data.repository.object === null) {
+        throw new InvalidParameter({
+          prettyMessage: 'package not found',
+        })
+      }
+      const entries = json.data.repository.object.entries
+      const directories = entries.filter(file => file.type === 'tree')
+      const versions = directories.map(file => file.name)
+      const version = latest(versions)
```

Comment:
Is there any way we could find out what the registry considers the "latest" version of a package? Do we have any info on how WinGet sorts versions?

I'm a bit worried that there doesn't really seem to be any constraint on what a version "number" can be. Here's some examples that have one or more odd ones:
- https://github.com/microsoft/winget-pkgs/tree/c939ca136ff33f217db1de99f1a1839790287a16/manifests/j/JetBrains/MPS/EAP
- https://github.com/microsoft/winget-pkgs/tree/c939ca136ff33f217db1de99f1a1839790287a16/manifests/n/niconico/nair/live
- https://github.com/microsoft/winget-pkgs/tree/c939ca136ff33f217db1de99f1a1839790287a16/manifests/j/Jigsaw/OutlineManager
- https://github.com/microsoft/winget-pkgs/tree/c939ca136ff33f217db1de99f1a1839790287a16/manifests/w/Wildy/bobonslaught
- https://github.com/microsoft/winget-pkgs/tree/c939ca136ff33f217db1de99f1a1839790287a16/manifests/t/taers232c/GAMADV-XTD3

Given it seems a version number can be basically any string, it seems like this we are going to get this wrong in some cases.

Reply:
Sorry for late response. I will investigate the update process in winget database

Reply:
I could noy find meaningful source code of winget-pkgs pipeline but I found version comparesion in winget-cli, that are used when one package is provided by multiple multiple sources.

I don't know if this compression is used by winget-pkgs pipeline and may not work well with some if your example packages but I think it's reasonable to use this comparesion.

I will implement this compression in services/winget folder with unit tests.

Version Implementation: https://github.com/microsoft/winget-cli/blob/ae566c7bf21cfcc75be7ec30e4036a30eede8396/src/AppInstallerSharedLib/Versions.cpp

Use case in winget-cli: https://github.com/microsoft/winget-cli/blob/ae566c7bf21cfcc75be7ec30e4036a30eede8396/src/AppInstallerRepositoryCore/PackageVersionSelection.cpp#L82-L100

Reply:
I have implemented in 6b707327b7bb6d0582d2c44ac32f1a6616ad5323

Reply:
Sorry  forget to impement latest case so nair case is my fault. 

however, I feel that JetBrains.MPS.EAP case is bug in winget registorysince actual latest version is 241.18034.990 and I think return it as latest with wingwt-cli version comparator.(I'll test later)

Reply:
I'd be down with that as a one-off check to verify this.
I'm not sure I have the appetite to write it.

Reply:
@anatawa12 , @PyvesB , @chris48s.

I made a script that dumps all the manifests to 

- `catalog.csv` - this contains packages and the versions
- `packages.csv` - contains the list of unique packages
- `packages-map.csv` - includes the version of the package returned by `winget show {packageId}` and the version returned by this PR. Last column shows if they match. Winget returns the version without the `v` prefix, in shields with the prefix. So I consider them a match if they have or not the `v` prefix.
- `packages-diff.csv` - contains those packages that need some your attention. 

N/A means it's not returned by `winget show` command (missing in the winget index but present in the manifests repo)

[catalog.csv](https://github.com/user-attachments/files/17532000/catalog.csv)
[packages.csv](https://github.com/user-attachments/files/17532002/packages.csv)
[packages-map.csv](https://github.com/user-attachments/files/17532015/packages-map.csv)
[packages-diff.csv](https://github.com/user-attachments/files/17532026/packages-diff.csv)

Hope this helps move this PR forward and eventually merge.

Reply:
Nice! 

So it looks like we've got 16 examples where our result disagrees with the upstream to look into there and it should be possible to use the lists of versions for those packages to write some test cases.

I'd hope that list of 16 examples reduces to fewer than 16 bugs.

@anatawa12 - are you happy to revisit this? I'm aware the PR has been dormant for some time at this point.

Reply:
Yes I want but It may took few days or weeks before working on this because I'm working on other projects now.

Reply:
I want to describe the cause of the problem for each package.

I fixed 12 of 16 but 4 are hard to fix I think

### Confusing part of package name and version name

Affected packages:
- `Canonical.Ubuntu`
- `Codemao.TurtleEditor`
- `ImageMagick.ImageMagick`
- `PLACE.IA`
- `PLACE.ONE`
- `qBittorrent.qBittorrent`
- `RoyalApps.RoyalTS`
- `Shure.WirelessWorkbench`
- `SoftMaker.FreePDF`

This is totally my fault.
I assumed such cases don't exist but it exists.

My code confuses `Canonical.Ubuntu.2404` package as `Canonical.Ubuntu` version 2024.
We can avoid this by checking if `<id>.yaml` exists in the version folder.

I fixed in aabbc388f773d677e1769df5497e1fdc762172e8.

### .0 version is incorrect

Affected packages:
- `CiderCollective.Cider.Nightly`

This is my fault.

In winget, version ends with `.0` will ignore `.0` but my code doesn't ignore `.0` correctly.

I fixed this in a58ab02f6c95ac1e36697dbb2e805afce9ca91e5.

### Same version in the registry

- `CodeF0x.violin` - `3.1.1` and `3.1.1.0`
- `UweSieber.UsbTreeView` - `4.4.2` and `4.4.2.0`

Those packages have same version according to the version comparator in winget.
Manifest content of those package versions are same.

I don't know how exactly winget handles this case, but it looks last entry in git repository wins, so I changed to use last entry in the repository in 439a7e0629db0da0e5c8cb7ea88d6de0f4b4d24e.

### Different package version comparator between 1.8 series and 1.9 series

Affected packages:
- `JetBrains.MPS.EAP`
- `maforget.ComicRackCE.Nightly`

I think this is not required to be fixed, and those packages are updated so It's impossible to test those cases.

It looks the version comparator in winget has been changed between 1.8 series and 1.9 series.
My implementation is based on 1.9 series and winget official repository is based on 1.8 series.

I have created https://github.com/anatawa12/winget-cli/tree/version-comparator-test (based on master, which is 1.9 series) and
cheery-picked onto v1.8.1791 (1.8 series) to compare the version comparator.
I confirmed our version matches with 1.9 series of winget.

For `JetBrains.MPS.EAP`, Preambles support added in [this commit](https://github.com/microsoft/winget-cli/commit/38aac878ab34d0176b34faf6946d7a0b55d4ce97) changed the version comparator.

For `maforget.ComicRackCE.Nightly`, also affected by Preambles support. `nightly-` part is treated as preamble and `7635` and `9` are compared in 1.9, but compared dictionary in 1.8.
Left is 1.9 series and right is 1.8 series. greater means newer, so greatest is the latest version. my implementation picks `nightly-7635d6e` as latest version, which is same as 1.9 series.

<img width="654" alt="image" src="https://github.com/user-attachments/assets/6ba4bf7b-dde1-4c77-b113-3b3209ee8138">

### Other failures

#### `Coninfer.ECLiPSeCLP.7.0` and `Coninfer.ECLiPSeCLP.7.1`

Showing those packages failed with the following error:
```
$ winget show --id='Coninfer.ECLiPSeCLP.7.1'
An unexpected error occurred while executing the command:
Download request status is not success.
0x80190194 : 見つかりません (404)。
```

Those packages have `#` in the version name.
as a part of `winget show` command, winget will request `https://cdn.winget.microsoft.com/cache/manifests/c/<package path>/<version>/<rowid in sqlite>` but `#` is not allowed in the URL.
I feel this is a bug in winget.

`winget show --id='<id>' --versions` shows the version my implementation picked, so I think it's OK to show the version `v7.0 #63` and `v7.1 #13` from shields.io. 

#### `1357310795.JboxTransfer`

I cannot `winget show --id=1357310795.JboxTransfer`. I don't know why it fails on winget side.

```
> winget show --id=1357310795.JboxTransfer
No package found matching input criteria.
```

Reply:
Thanks for laying all this out really clearly.

I'm inclined to agree about the handful of cases where we don't need to be "bug for bug" compatible

Reply:
@chris48s  any chance we can merge this?